src/actions/hover.rs

```diff
@@ -0,0 +1,2067 @@
+            In hac habitasse platea dictumst.
+
+            End module docs.
+        ",
+        );
+
+        assert_eq!(expected, actual, "module docs without a copyright header");
+    }
+
+    #[test]
+    fn test_extract_docs_comment_block() {
+        let vfs = Vfs::new();
+        let file = Path::new("test_data/hover/src/test_extract_docs_comment_block.rs");
+        let row_start = Row::new_zero_indexed(21);
+        let actual = extract_docs(&vfs, &file, row_start)
+            .expect(&format!("failed to extract docs: {:?}", file))
+            .join("\n");
+
+        let expected = noindent(
+            "
+            The standard library often has comment header blocks that should not be
+            included.
+
+            Nam efficitur dapibus lectus consequat porta. Pellentesque augue metus,
+            vestibulum nec massa at, aliquet consequat ex.
+
+            End of spawn docs
+        ",
+        );
+
+        assert_eq!(expected, actual);
+    }
+
+    #[test]
+    fn test_extract_docs_empty_line_before_decl() {
+        let vfs = Vfs::new();
+        let file = Path::new("test_data/hover/src/test_extract_docs_empty_line_before_decl.rs");
+        let row_start = Row::new_zero_indexed(18);
+        let actual = extract_docs(&vfs, &file, row_start)
+            .expect(&format!("failed to extract docs: {:?}", file))
+            .join("\n");
+
+        let expected = noindent(
+            "
+            Begin empty before decl
+
+            Cras malesuada mattis massa quis ornare. Suspendisse in ex maximus,
+            iaculis ante non, ultricies nulla. Nam ultrices convallis ex, vel
+            lacinia est rhoncus sed. Nullam sollicitudin finibus ex at placerat.
+
+            End empty line before decl.
+        ",
+        );
+
+        assert_eq!(expected, actual);
+    }
+
+    #[test]
+    fn test_extract_docs_module_docs() {
+        let vfs = Vfs::new();
+        let file = Path::new("test_data/hover/src/test_extract_docs_module_docs.rs");
+
+        let row_start = Row::new_zero_indexed(0);
+        let actual = extract_docs(&vfs, &file, row_start)
+            .expect(&format!("failed to extract docs: {:?}", file))
+            .join("\n");
+
+        let expected = noindent(
+            "
+            Begin module docs
+
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas
+            tincidunt tristique maximus. Sed venenatis urna vel sagittis tempus.
+            In hac habitasse platea dictumst.
+
+            End module docs.
+        ",
+        );
+
+        assert_eq!(expected, actual);
+
+        let row_start = Row::new_zero_indexed(21);
+        let actual = extract_docs(&vfs, &file, row_start)
+            .expect(&format!("failed to extract docs: {:?}", file))
+            .join("\n");
+
+        let expected = noindent(
+            "
+            Begin first item docs
+
+            The first item docs should not pick
```

Comment:
It's great to see such an extensive test coverage, that's an awesome thing to have!

While I certainly don't mind having more tests, I'm somewhat concerned this blows up number of currently used test solutions - with this, we'll have integration tests in a form of Cargo unit tests, only a couple of actual Cargo-style integration tests, [additional json tests for diagnostics](https://github.com/rust-lang-nursery/rls/blob/master/src/actions/diagnostics.rs#L341) and now this custom harness for tooltips (which, with test definitions, takes a bit more LOC than the actual logic itself).

It shouldn't block this PR and it's a great thing to have but just as a note, I think we need to take a more streamlined and declarative approach to testing various RLS bits in the future.